ddp-workspace/projects/ddp-singular/src/app/app.component.ts

```diff
@@ -1,18 +1,52 @@
-import { Component, ElementRef } from '@angular/core';
+import { Component, ElementRef, HostListener } from '@angular/core';
+import { AnalyticsEventsService } from 'ddp-sdk';
+
+import { GTagEvent } from './constants/gtag-event';
 
 
 @Component({
   selector: 'app-root',
   templateUrl: './app.component.html',
   styleUrls: ['./app.component.scss']
 })
-export class AppComponent  {
+export class AppComponent {
   title = 'ddp-singular';
 
-  constructor(private elRef: ElementRef) {
-  }
+  constructor(
+    private elRef: ElementRef,
+    private analytics: AnalyticsEventsService
+  ) {}
 
   onActivate(): void {
     this.elRef.nativeElement.scrollTo(0,0);
   }
+
+  // based on https://plausible.io/blog/track-outbound-link-clicks#:~:text=%3C-,script,-%3E%0A%20%20document
+  @HostListener('document:click', ['$event'])
+  monitorOutboundLinks(event: Event): void {
+    const link = this.findNearestAnchorElement(event.target);
+
+    // stop if it's the same origin navigation
+    if (!link || !link.href || !link.host || link.host === location.host) {
+      return;
+    }
+
+    this.analytics.emitCustomGtagEvent(GTagEvent.OUTBOUND_LINK_CLICK, link.innerText, link.href);
+
+    // Allow event to be sent before the page is unloaded, if we're navigating in the same page
+    if (!link.target || link.target.match(/^_(self|parent|top)$/i)) {
+      setTimeout(function() { location.href = link.href; }, 100);
+      event.preventDefault();
+    }
+  }
```

Comment:
I think the hostlistener is really nice idea!
But we might need to tweak things to avoid duplicate event generation.
e.g.: noticing that when the "preview consent form" or "preview kit instructions" are clicked, we are generating 2 gtag entries: the one I created in code for pdf_downloads and the one driven by the hostlistener execution. 

Same for Facebook clicks (note the 2 highlighted "en" entries in decode gtag.js payload:
![Screen Shot 2022-05-15 at 11 19 05 AM](https://user-images.githubusercontent.com/29984380/168480285-1c10309d-d06b-49f7-90a7-b2bd169ff585.png)
:

Reply:
On the duplicate Facebook entries: it occurs to me that this MIGHT be OK; it's just another outbound link so there is it's event. But if we are going to duplicate events, I think we want to be intentional about it (as opposed to an unexpected consequence of our code).

Reply:
@grigol-brd Regarding the failed tests you mentioned on Slack: I just removed them. They were just the auto-generated tests for a couple of the static pages. All tests pass on this branch now.

Reply:
@MocanaAtBroad yea, I noticed that but wasn't sure how much of an issue that is, thanks for bringing it up.
I'll take some time to try to find a good solution to this, but just in case, how can we be intentional about it? Can you elaborate on what you mean?

Reply:
I just meant that it could very well be that it does not matter that we are generating duplicate events IF the consumer of the data is able (hopefully easily) to filter data in a useful. We just don't know though, so we should try to provide as clean data as we can.

A possible solution that would be quick and easy: tweak your hostlistener function to look for anchors AND that have a CSS class name (e.g., "analytics"). CSS class has no style associated with it, just used to query for or filter the links that should generate a gtag.js event.

Reply:
@MocanaAtBroad I see, yep your suggested solution seems simpler, only if I saw this reply earlier : (
The way I did it is to attach metadata to the event object upon gtag handling so when it propagates up the topmost host listener will only proceed with unhandled ones.

Reply:
@MocanaAtBroad right, I changed it to use CSS class